cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go

```diff
@@ -662,16 +678,91 @@ func (c *client) WaitForResourceStatuses() error {
-			// 	return false, nil
-			// }
 		}
 
 		return true, nil
 	})
 }
 
+func (c *client) GetClusterSecrets(cluster *clusterv1.Cluster) ([]*corev1.Secret, error) {
+	list := &corev1.SecretList{}
+	if err := c.clientSet.List(ctx, list, ctrlclient.InNamespace(cluster.Namespace)); err != nil {
+		return nil, errors.Wrapf(err, "error listing Secrets for Cluster %s/%s", cluster.Namespace, cluster.Name)
+	}
+
+	res := []*corev1.Secret{}
+	for i, secret := range list.Items {
+		if strings.HasPrefix(secret.Name, cluster.Name) {
+			res = append(res, &list.Items[i])
+		}
+	}
+	return res, nil
+}
+
+func (c *client) CreateSecret(s *corev1.Secret) error {
+	if err := c.clientSet.Create(context.Background(), s); err != nil {
+		return errors.Wrapf(err, "error creating Secret %s/%s", s.Namespace, s.Name)
+	}
+	return nil
+}
+
+func (c *client) ForceDeleteSecret(namespace, name string) error {
+	secret := &corev1.Secret{}
+	if err := c.clientSet.Get(ctx, ctrlclient.ObjectKey{Namespace: namespace, Name: name}, secret); err != nil {
+		return errors.Wrapf(err, "error getting Secret %s/%s", namespace, name)
+	}
+
+	if len(secret.Finalizers) > 0 {
+		secret.SetFinalizers([]string{})
+		if err := c.clientSet.Update(ctx, secret); err != nil {
+			return errors.Wrapf(err, "error removing finalizer for Secret %s/%s", secret.Namespace, secret.Name)
+		}
+	}
+
+	if err := c.clientSet.Delete(context.Background(), secret); err != nil {
+		return errors.Wrapf(err, "error deleting Secret %s/%s", secret.Namespace, secret.Name)
+	}
+	return nil
+}
+
+func (c *client) GetUnstructuredObject(u *unstructured.Unstructured) error {
+	key := ctrlclient.ObjectKey{Namespace: u.GetNamespace(), Name: u.GetName()}
+	if err := c.clientSet.Get(context.Background(), key, u); err != nil {
+		return errors.Wrapf(err, "error fetching unstructured object %q %v", u.GroupVersionKind(), key)
+	}
+	return nil
+}
+
+func (c *client) CreateUnstructuredObject(u *unstructured.Unstructured) error {
+	if err := c.clientSet.Create(context.Background(), u); err != nil {
+		return errors.Wrapf(err, "error creating unstructured object %q %s/%s",
+			u.GroupVersionKind(), u.GetNamespace(), u.GetName())
+	}
+	return nil
+}
+
+// DeleteUnstructuredObjects deletes all the unstructured objects of the same type in a namespace.
+// If the namespace is empty then all Clusters in all namespaces are deleted.
+func (c *client) DeleteUnstructuredObjects(namespace string, u *unstructured.Unstructured) error {
+	if err := c.clientSet.DeleteAllOf(ctx, u, ctrlclient.InNamespace(namespace)); err != nil {
+		return errors.Wrapf(err, "error deleting unstructured objects %q in namespace %q", u.GroupVersionKind(), namespace)
+	}
+	return nil
+}
+
+func (c *client) ForceDeleteUnstructuredObject(u *unstructured.Unstructured) error {
+	u.SetFinalizers([]string{})
+	if err := c.clientSet.Update(ctx, u); err != nil {
+		return errors.Wrapf(err, "error removing finalizer for unstructured object %q %s/%s",
```

Comment:
In running this locally I'm hitting a handful of errors in the force deletion:

```
F0822 10:44:01.693707   99289 create_cluster.go:59] unable to pivot cluster api stack to target cluster: unable to pivot cluster API objects: failed to move Machine default:rudeboy-pizzatown-control-0: error copying Machine default/rudeboy-pizzatown-control-0 bootstrap reference to target cluster: error force deleting unstructured object "bootstrap.cluster.x-k8s.io/v1alpha2, Kind=KubeadmConfig" default/rudeboy-pizzatown-config to target cluster: error removing finalizer for unstructured object "bootstrap.cluster.x-k8s.io/v1alpha2, Kind=KubeadmConfig" default/rudeboy-pizzatown-config: Operation cannot be fulfilled on kubeadmconfigs.bootstrap.cluster.x-k8s.io "rudeboy-pizzatown-config": StorageError: invalid object, Code: 4, Key: /registry/bootstrap.cluster.x-k8s.io/kubeadmconfigs/default/rudeboy-pizzatown-config, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: b79e65c2-8416-41c0-8d20-a2399de8bbb3, UID in object meta: 
```

Would it be worth it to just bypass any IsNotFound errors or errors like the one above which only occur when the resource is gone? Or perhaps add a Get before trying to patch the finalizers?

Reply:
That's a good point, it'll ignore the not found errors

Reply:
Fixed, ptal :)

Reply:
Hmm so sometimes the error that comes back is `Precondition failed: UID in precondition: b79e65c2-8416-41c0-8d20-a2399de8bbb3, UID in object meta: `, which I think maps to `apierrors.IsInvalid()`, and I'm not sure if it would be right to bypass that type...

Reply:
This could totally be the result of mistakes in my own YAML so I'm not sure I wanna block the PR on it, but, for example, my pivot did work if I ignored `apierrors.IsInvalid()`. Confusing to say the least

Reply:
Paired with @vincepri for a bit and this is fixed up now. :)